### DISCUSSION TITLE
Bulk Endpoint Help

### DISCUSSION THREAD

**Discussion by 641i130** [hathach/tinyusb#1878] [Q&A]
I'm trying to write a basic bulk endpoint program (emulating an arcade IO board). I've documented the [protocol here](https://github.com/641i130/ttx4-usbio) and the project I'm working that needs a bulk endpoint [is here](https://github.com/641i130/pico-ttx4). I have everything matching via a verbose lsusb. I'm just very confused on how to create and manage a bulk endpoint.

**Comment by hathach** [comment]
which part of bulk endpoint is confusing ?

**Reply by 641i130** [reply]
I'm unsure how I should be going about making it accessible. Maybe my implementation is wrong, here is my code:
https://github.com/641i130/pico-ttx4/blob/main/src/main.c

I haven't worked much with USB, but I did end up getting that USB textbook you recommended in my other help requests in the past, read through the portions on bulk but still don't fully understand it.

**Reply by hathach** [reply]
there is nothing special about Bulk, it works like other endpoint type like Interrupt. But you only need to bind/claim the interface or set udev to grant OS permission to make the transfer. You can use the vendor class driver, which will populate 2 bulk endpoints in & out for usage. There is a webusb example that make use of the bulk endpoints

**Reply by 641i130** [reply]
I guess I'm missing that portion of the code. Thank you, I'll look into this.
